### PR TITLE
Fields field in security rule takes values from args object

### LINKED ISSUE #1270: The fields section in remove, encrypt, decrypt and hash can be a variable
## The problem faced currently?
Currently the `fields` field in the security rule object can only be an array of string. There is no mechanism to load this from the `args` object.
### How can we solve it?
We should make the `fields` an `interface{}` instead of `[]string`.
>If you want this feature to be implemented, give it a thumbs up reaction, so that we can determine which features are important to you.
>👍

### PR DESCRIPTION
**Pull request by sharadregoti**
Fixes #1270

### CODE REVIEW

File: gateway/modules/auth/operations.go [hunk 1]
```diff
@@ -12,3 +17,22 @@ func (m *Module) Encrypt(value string) (string, error) {
 
 	return utils.Encrypt(m.aesKey, value)
 }
+
+func (m *Module) getFields(ctx context.Context, fields interface{}, args map[string]interface{}) ([]interface{}, error) {
```
Comment (YourTechBud) [inline gateway/modules/auth/operations.go:21]: Can we return array of string here?

Comment (YourTechBud) [reply gateway/modules/auth/operations.go:21]: I suggest you move this function to `helpers.go`

**Review by YourTechBud — CHANGES_REQUESTED** [review]
Resolve the conflicts as well

File: gateway/modules/auth/operations.go (hunk 1 above)
Comment (sharadregoti) [reply gateway/modules/auth/operations.go:21]: [] string? we can but an unnecessary for loop

Comment (sharadregoti) [reply gateway/modules/auth/operations.go:21]: There is no helpers.go in auth package?

Comment (YourTechBud) [reply gateway/modules/auth/operations.go:21]: So lets return [] interface only. Also, make a helpers.go file

File: gateway/modules/auth/match.go [hunk 2]
```diff
@@ -201,7 +201,15 @@ func (m *Module) matchRemove(ctx context.Context, projectID string, rule *config
 		}
 	}
 	actions := &model.PostProcess{}
-	for _, field := range rule.Fields {
+	fields, err := m.getFields(ctx, rule.Fields, args)
+	if err != nil {
+		return nil, helpers.Logger.LogError(helpers.GetRequestID(ctx), "Invalid value provided for field (Fields) in security rules where rule is (Remove)", err, nil)
```
Comment (YourTechBud) [inline gateway/modules/auth/match.go:206]:
```suggestion
		return nil, err
```

File: gateway/modules/auth/match.go [hunk 3]
```diff
@@ -228,7 +236,15 @@ func (m *Module) matchEncrypt(ctx context.Context, projectID string, rule *confi
 		}
 	}
 
-	for _, field := range rule.Fields {
+	fields, err := m.getFields(ctx, rule.Fields, args)
+	if err != nil {
+		return nil, helpers.Logger.LogError(helpers.GetRequestID(ctx), "Invalid value provided for field (Fields) in security rules where rule is (Encrypt)", err, nil)
```
Comment (YourTechBud) [inline gateway/modules/auth/match.go:241]:
```suggestion
		return nil, err
```

File: gateway/modules/auth/match.go [hunk 4]
```diff
@@ -267,7 +282,15 @@ func (m *Module) matchDecrypt(ctx context.Context, projectID string, rule *confi
 		}
 	}
 
-	for _, field := range rule.Fields {
+	fields, err := m.getFields(ctx, rule.Fields, args)
+	if err != nil {
+		return nil, helpers.Logger.LogError(helpers.GetRequestID(ctx), "Invalid value provided for field (Fields) in security rules where rule is (Decrypt)", err, nil)
```
Comment (YourTechBud) [inline gateway/modules/auth/match.go:287]:
```suggestion
		return nil, err
```

File: gateway/modules/auth/match.go [hunk 5]
```diff
@@ -321,7 +343,15 @@ func (m *Module) matchHash(ctx context.Context, projectID string, rule *config.R
 		}
 	}
 
-	for _, field := range rule.Fields {
+	fields, err := m.getFields(ctx, rule.Fields, args)
+	if err != nil {
+		return nil, helpers.Logger.LogError(helpers.GetRequestID(ctx), "Invalid value provided for field (Fields) in security rules where rule is (Hash)", err, nil)
```
Comment (YourTechBud) [inline gateway/modules/auth/match.go:348]:
```suggestion
		return nil, err
```

**Review by YourTechBud — CHANGES_REQUESTED** [review]

**Review by YourTechBud — APPROVED** [review]

**Pull request merged**